Features/buildrel (#402)
dotnet/machinelearning-samples: samples/csharp/getting-started/AdvancedExperiment_AutoML/README.md

# Taxi Fare Prediction - Advanced AutoML Experiment

## Automated Machine Learning

> Automated machine learning (AutoML) automates the end-to-end process of applying machine learning to real-world problems. Given a dataset, AutoML iterates over different data featurizations, machine learning algorithms, hyperparamters, etc. to select the best model based on training scores. 

Reviewer: We'll hopefully be choosing on more criteria besides the training scores, like training speed, model size, prediction speed, memory, etc in the future...
```suggestion
Automated machine learning (AutoML) automates the end-to-end process of applying machine learning to real-world problems. Given a dataset, AutoML iterates over different data featurizations, machine learning algorithms, hyperparamters, etc. to select the best model for the dataset. 
```

Reviewer 2: Ah, thanks! (Nit: since the sentence starts with 'Given a dataset', going to just to end sentence after 'model')

```C#            
 ExperimentResult<RegressionMetrics> experimentResult = experiment.Execute(TrainSmallDataView, columnInformation, preFeaturizer, progressHandler);
```

## Step 8: Re-fit Best Pipeline
Re-fit the best pipeline (trained from subsample of AutoML data) on entirety of training data. (This step is optional. By no means is it required in your workflow.)

```C#
private static ITransformer RefitBestPipeline(MLContext mlContext, ExperimentResult<RegressionMetrics> experimentResult)
{
    RunDetail<RegressionMetrics> best = experimentResult.BestRun;
    return best.Estimator.Fit(TrainDataView);
}
```

## Step 9: Evaluate Model

Evaluate the quality of the re-fit model on a test dataset that was not used in training (`taxi-fare-test.csv`).

```C#
IDataView predictions = model.Transform(TestDataView);
var metrics = mlContext.Regression.Evaluate(predictions, labelColumnName: LabelColumnName, scoreColumnName: "Score");
```

> ## Step 10: Make Predictions

Reviewer: There should be a refit on the entire dataset (Train+Test). Metrics are produced in step 9, the actual model to send to production is created in this final refit.

Reviewer 2: Moved re-fit for to the end in this advanced sample that re-fits on train & test, and then saves that model for use in production
